api/src/main/kotlin/com/github/benmanes/gradle/versions/reporter/PlainTextReporter.kt

```diff
@@ -0,0 +1,194 @@
+  override val project: Project,
+  override val revision: String,
+  override val gradleReleaseChannel: String,
+) : AbstractReporter(project, revision, gradleReleaseChannel) {
+  /** Writes the report to the print stream. The stream is not automatically closed. */
+  override fun write(printStream: Appendable, result: Result) {
+    writeHeader(printStream)
+
+    if (result.count == 0) {
+      printStream.println()
+      printStream.println("No dependencies found.")
+    } else {
+      writeUpToDate(printStream, result)
+      writeExceedLatestFound(printStream, result)
+      writeUpgrades(printStream, result)
+      writeUndeclared(printStream, result)
+      writeUnresolved(printStream, result)
+    }
+
+    writeGradleUpdates(printStream, result)
+  }
+
+  override fun getFileExtension(): String {
+    return "txt"
+  }
+
+  private fun writeHeader(printStream: Appendable) {
+    printStream.println()
+    printStream.println("------------------------------------------------------------")
+    printStream.println("${project.path} Project Dependency Updates (report to plain text file)")
+    printStream.println("------------------------------------------------------------")
+  }
+
+  private fun writeUpToDate(printStream: Appendable, result: Result) {
+    val upToDateVersions = result.current.dependencies
+    if (upToDateVersions.isNotEmpty()) {
+      printStream.println()
+      printStream.println("The following dependencies are using the latest $revision version:")
+      for (dependency in upToDateVersions) {
+        printStream.println(" - ${label(dependency)}:${dependency.version}")
+        dependency.userReason?.let {
+          printStream.println("     $it")
+        }
+      }
+    }
+  }
+
+  private fun writeExceedLatestFound(printStream: Appendable, result: Result) {
+    val downgradeVersions = result.exceeded.dependencies
+    if (downgradeVersions.isNotEmpty()) {
+      printStream.println()
+      printStream.println(
+        "The following dependencies exceed the version found at the $revision revision level:"
+      )
+      for (dependency in downgradeVersions) {
+        val currentVersion = dependency.version
+        printStream.println(" - ${label(dependency)} [${currentVersion} <- ${dependency.latest}]")
+        dependency.userReason?.let {
+          printStream.println("     $it")
+        }
+        dependency.projectUrl?.let {
+          printStream.println("     $it")
+        }
+      }
+    }
+  }
+
+  private fun writeUpgrades(printStream: Appendable, result: Result) {
+    val upgradeVersions = result.outdated.dependencies
+    if (upgradeVersions.isNotEmpty()) {
+      printStream.println()
+      printStream.println("The following dependencies have later $revision versions:")
+      for (dependency in upgradeVersions) {
+        val currentVersion = dependency.version
+//        printStream.println(" - ${label(dep)} [${currentVersion} -> ${dep.available[revision]}]")
```

Comment:
?

Reply:
Groovy:
```
  private void writeUpgrades(Appendable printStream, Result result) {
    Set<DependencyOutdated> upgradeVersions = result.outdated.dependencies
    if (!upgradeVersions.isEmpty()) {
      printStream.println()
      printStream.println("The following dependencies have later ${revision} versions:")
      for (DependencyOutdated dep in upgradeVersions) {
        String currentVersion = dep.version
        printStream.println(" - ${label(dep)} [${currentVersion} -> ${dep.available[revision]}]")
        if (dep.getUserReason()) {
          printStream.println("     ${dep.getUserReason()}")
        }
        if (dep.projectUrl != null) {
          printStream.println("     ${dep.projectUrl}")
        }
      }
    }
  }
```

Kotlin:
```
  private fun writeUpgrades(printStream: Appendable, result: Result) {
    val upgradeVersions = result.outdated.dependencies
    if (upgradeVersions.isNotEmpty()) {
      printStream.println()
      printStream.println("The following dependencies have later $revision versions:")
      for (dependency in upgradeVersions) {
        val currentVersion = dependency.version
//        printStream.println(" - ${label(dep)} [${currentVersion} -> ${dep.available[revision]}]")
        dependency.userReason?.let {
          printStream.println("     $it")
        }
        dependency.projectUrl?.let {
          printStream.println("     $it")
        }
      }
    }
  }
```

I am not sure where `${dep.available[revision]}` resolves to.

Reply:
It’s a lookup into this
https://github.com/ben-manes/gradle-versions-plugin/blob/df508c6b4b9cf096958bd1157b92d208b1e342c4/api/src/main/kotlin/com/github/benmanes/gradle/versions/reporter/result/DependencyOutdated.kt#L9

Reply:
Spoke offline.